File: examples/simple_server.c
/*
 * Copyright (c) 2004-2010 The Trustees of Indiana University and Indiana
 *                         University Research and Technology
 *                         Corporation.  All rights reserved.
 * Copyright (c) 2004-2011 The University of Tennessee and The University
 *                         of Tennessee Research Foundation.  All rights
 *                         reserved.
 * Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
 *                         University of Stuttgart.  All rights reserved.
 * Copyright (c) 2004-2005 The Regents of the University of California.
 *                         All rights reserved.
 * Copyright (c) 2006-2013 Los Alamos National Security, LLC.
 *                         All rights reserved.
 * Copyright (c) 2009-2012 Cisco Systems, Inc.  All rights reserved.
 * Copyright (c) 2011      Oak Ridge National Labs.  All rights reserved.
 * Copyright (c) 2013-2019 Intel, Inc.  All rights reserved.
 * Copyright (c) 2015      Mellanox Technologies, Inc.  All rights reserved.
 * Copyright (c) 2021-2026 Nanook Consulting  All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 *
 */

#define _GNU_SOURCE
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "examples.h"
#include <pmix.h>
#include <pmix_server.h>

static pmix_server_module_t mymodule = {
    .client_connected = NULL,
    .client_finalized = NULL,
    .abort = NULL,
    .fence_nb = NULL,
    .direct_modex = NULL,
    .publish = NULL,
    .lookup = NULL,
    .unpublish = NULL,
    .spawn = NULL,
    .connect = NULL,
    .disconnect = NULL,
    .register_events = NULL,
    .deregister_events = NULL,
    .notify_event = NULL,
    .query = NULL,
    .tool_connected = NULL,
    .log = NULL,
    .allocate = NULL,
    .job_control = NULL,
    .monitor = NULL,
    .group = NULL
};


int main(int argc, char **argv)
{
    pmix_status_t rc;
    EXAMPLES_HIDE_UNUSED_PARAMS(argc, argv);

    /* init us - note that the call to "init" includes the return of
     * any job-related info provided by the RM. This includes any
     * debugger flag instructing us to stop-in-init. If such a directive
     * is included, then the process will be stopped in this call until
     * the "debugger release" notification arrives */
    if (PMIX_SUCCESS != (rc = PMIx_server_init(&mymodule, NULL, 0))) {
        fprintf(stderr, "Server_init failed: %s\n", PMIx_Error_string(rc));
        exit(1);
    }

    rc = PMIx_server_finalize();
    if (PMIX_SUCCESS != rc) {
        fprintf(stderr, "Server finalize failed: %s\n", PMIx_Error_string(rc));
    }

    fprintf(stderr, "Pass one completed\n");

    if (PMIX_SUCCESS != (rc = PMIx_server_init(&mymodule, NULL, 0))) {
        fprintf(stderr, "Server_init second time failed: %s\n", PMIx_Error_string(rc));
        exit(1);
    }

   /* finalize us */
    rc = PMIx_server_finalize();
    if (PMIX_SUCCESS != rc) {
        fprintf(stderr, "Server finalize second time failed: %s\n", PMIx_Error_string(rc));
    }

    fprintf(stderr, "Pass two completed\n");

    fflush(stderr);
    return (rc);
}

File: examples/python/simple_server.py
#!/usr/bin/env python3
#
# Copyright (c) 2026      Nanook Consulting  All rights reserved.
# $COPYRIGHT$
#
# Additional copyrights may follow
#
# $HEADER$
#
# Python port of examples/simple_server.c
#
# Bring a PMIx server up and down twice.
#
# One deviation from the C version: it declares a server module whose
# every entry is NULL, which the C API accepts. The Python binding
# deliberately refuses an empty module map ("SERVER REQUIRES AT LEAST ONE
# MODULE FUNCTION TO OPERATE"), so this port registers a single trivial
# handler instead. What the example is really exercising - that a server
# can be initialized and finalized repeatedly - is unchanged.

import sys

from examples import *


def clientconnected(proc, args):
    """The one module function we register, so init() has something."""
    return PMIX_SUCCESS


def main():
    server = PMIxServer()
    mymodule = {'clientconnected': clientconnected}

    # init us - note that the call to "init" includes the return of
    # any job-related info provided by the RM. This includes any
    # debugger flag instructing us to stop-in-init. If such a directive
    # is included, then the process will be stopped in this call until
    # the "debugger release" notification arrives
    rc = server.init(None, mymodule)
    if PMIX_SUCCESS != rc:
        eprint("Server_init failed:", server.error_string(rc))
        sys.exit(1)

    rc = server.finalize()
    if PMIX_SUCCESS != rc:
        eprint("Server finalize failed:", server.error_string(rc))

    eprint("Pass one completed")

    rc = server.init(None, mymodule)
    if PMIX_SUCCESS != rc:
        eprint("Server_init second time failed:", server.error_string(rc))
        sys.exit(1)

    # finalize us
    rc = server.finalize()
    if PMIX_SUCCESS != rc:
        eprint("Server finalize second time failed:",
               server.error_string(rc))

    eprint("Pass two completed")

    return rc


if __name__ == '__main__':
    sys.exit(main())
